### DISCUSSION TITLE
Native AOT: Microsoft.Extensions.DependencyInjection

### DISCUSSION THREAD

**Discussion by weifenluo** [dotnet/runtime#110386] [Q&A] [answered]
Is `Microsoft.Extensions.DependencyInjection` native AOT compatible? I assume it uses reflection for DI intensively, and I can't seem to find a source generator mode like provided by `System.Text.Json`.

I'm asking this question because I'm evaluating whether I should build my library on top of it without losing native AOT compatibility.

**Comment by davidfowl** [comment, marked as answer]
Yea, it's AOT compatible, there's no source generator for it. See https://github.com/dotnet/runtime/discussions/91675#discussioncomment-9783240 for more details.

**Comment by weifenluo** [comment]
@davidfowl  Thanks for your reply, it's very informative!

I still don't understand how `Microsoft.Extensions.DependencyInjection` can be AOT compatible without any source generator. Given the following EF Core DbContext constructor:
```
public class MyDbContext(DbContextOptions<MyDbContext> options) : DbContext
    : base(options)
{
    ...
}
``` 
There is nowhere else calling this constructor except DI via reflection. How can we prevent this constructor from being trimmed?

**Reply by davidfowl** [reply]
MEDI uses annotations to root services registered. Methods like AddDbContext [are annotated](https://github.com/dotnet/efcore/blob/ac5bd7bc664b44837c515da14577f4c48a63849b/src/EFCore/Extensions/EntityFrameworkServiceCollectionExtensions.cs#L60) with the right set of attributes to let the trimmer know what to keep when using DI.

Read more about it here https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/prepare-libraries-for-trimming
Watch the video here https://youtu.be/N-MrQeZ1enY?si=g3Fv2HLdKmtlUTVD

PS: If you don't get AOT warnings when you compile, then it should work just fine at runtime.
